imap_processing/ultra/l1c/ultra_l1c_pset_bins.py

```diff
@@ -65,124 +73,71 @@ def build_spatial_bins(
-    """
-    vx = v[:, 0]
-    vy = v[:, 1]
-    vz = v[:, 2]
-
-    # Magnitude of the velocity vector
-    magnitude_v = np.sqrt(vx**2 + vy**2 + vz**2)
-
-    vhat_x = -vx / magnitude_v
-    vhat_y = -vy / magnitude_v
-    vhat_z = -vz / magnitude_v
-
-    # Elevation angle (angle from the z-axis, range: [-pi/2, pi/2])
-    el = np.arcsin(vhat_z)
-
-    # Azimuth angle (angle in the xy-plane, range: [0, 2*pi])
-    az = np.arctan2(vhat_y, vhat_x)
-
-    # Ensure azimuth is from 0 to 2PI
-    az = az % (2 * np.pi)
-
-    return np.degrees(az), np.degrees(el), magnitude_v
-
-
-def spherical_to_cartesian(
-    r: np.ndarray, theta: np.ndarray, phi: np.ndarray
-) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
-    """
-    Convert spherical coordinates to Cartesian coordinates.
-
-    Parameters
-    ----------
-    r : np.ndarray
-        Radius.
-    theta : np.ndarray
-        Azimuth angle in radians.
-    phi : array-like or float
-        Elevation angle in radians.
-
-    Returns
-    -------
-    x, y, z : tuple
-        Cartesian coordinates.
-    """
-    x = r * np.cos(phi) * np.cos(theta)
-    y = r * np.cos(phi) * np.sin(theta)
-    z = r * np.sin(phi)
-
-    return x, y, z
-
-
 def get_histogram(
-    v: tuple[np.ndarray, np.ndarray, np.ndarray],
+    vhat: tuple[np.ndarray, np.ndarray, np.ndarray],
     energy: np.ndarray,
     az_bin_edges: np.ndarray,
     el_bin_edges: np.ndarray,
-    energy_bin_edges: np.ndarray,
+    energy_bin_edges: list[tuple[float, float]],
 ) -> NDArray:
     """
     Compute a 3D histogram of the particle data.
 
     Parameters
     ----------
-    v : tuple[np.ndarray, np.ndarray, np.ndarray]
-        The x,y,z-components of the velocity vector.
+    vhat : tuple[np.ndarray, np.ndarray, np.ndarray]
+        The x,y,z-components of the unit velocity vector.
     energy : np.ndarray
         The particle energy.
     az_bin_edges : np.ndarray
         Array of azimuth bin boundary values.
     el_bin_edges : np.ndarray
         Array of elevation bin boundary values.
-    energy_bin_edges : np.ndarray
+    energy_bin_edges : list[tuple[float, float]]
         Array of energy bin edges.
 
     Returns
     -------
     hist : np.ndarray
         A 3D histogram array.
+
+    Notes
+    -----
+    The histogram will now work properly for overlapping energy bins, i.e.
+    the same energy value can fall into multiple bins if the intervals overlap.
     """
-    az, el, _ = cartesian_to_spherical(v)
+    spherical_coords = cartesian_to_spherical(vhat)
+    az, el = (
+        spherical_coords[..., 1],
+        spherical_coords[..., 2],
+    )
 
-    # 3D binning.
-    hist, _ = np.histogramdd(
-        sample=(az, el, energy), bins=[az_bin_edges, el_bin_edges, energy_bin_edges]
+    # Initialize histogram
+    hist_total = np.zeros(
+        (len(az_bin_edges) - 1, len(el_bin_edges) - 1, len(energy_bin_edges))
     )
 
-    return hist
+    for i, (e_min, e_max) in enumerate(energy_bin_edges):
```

Comment:
Are the energy bins such that they can't be passed directly to `np.histogramdd()` if they were made to be an array of edges like `az_bin_edges`? Like:
```
hist_total = histogramdd(
    (az.ravel(), el.ravel(), energy.ravel()),
    bins=[az_bin_edges, el_bin_edges, energy_bin_edges]
)
```
If there are no gaps between the energy bands, it seems like `energy_bin_edges` could be made to be an array of edges rather than a list of tuples.

Reply:
That is how I had it before, but the Ultra team told me that there are **overlapping** energy bins so I had to make some changes.